### PR TITLE
Port multiserver-dht to Hyperswarm

### PR DESCRIPTION
**Pull request by KyleMaas**
My attempt at porting to Hyperswarm.  (I don't have a second computer handy to try to test this from, so I'm hoping this works.)

I changed example-s.js from using "keys" to using a single "key" because there is an issue (present in master as well) where it actually spins up servers on each *letter* of each key rather than each key as a whole.  That will need to be fixed, too, but in the meantime this change should allow for proper testing in the method that's most normally used.

### CODE REVIEW

**Comment by KyleMaas** [comment]
Oh, shoot, I just saw your "hyperswarm" branch.  Why wasn't that ever used?

**Comment by KyleMaas** [comment]
Got another computer going and was able to test it.  The test scripts needed to use the same key on either end, but otherwise it works quite well.  It's very quick and seems to work pretty easily.

**Comment by KyleMaas** [comment]
This should now work both on the command line and in the browser via hyperswarm-web and a browser switch in package.json.

File: index.js [hunk 1]
```diff
@@ -43,20 +39,15 @@ function updateChannelsToHost(onError, serverCfg) {
...
+        serverCfg.peer.join(crypto.createHash('sha256').update(channel).digest(), { lookup: true, announce: true });
       }
     });
 
     // oldChannels minus newChannels => leave
     oldChannels.forEach(channel => {
       if (!newChannels.has(channel)) {
         serverCfg.channels.delete(channel);
-        serverCfg.peer.leave(channel);
+        serverCfg.peer.leave(crypto.createHash('sha256').update(channel).digest());
```
Comment (staltz) [inline index.js:50]: This is used quite some times, could you make a function `channelToBuf`?

File: index.js [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
 function createPeer(_opts) {
   var swarmOpts = Object.assign({}, _opts || {});
   delete swarmOpts.key;
   delete swarmOpts.keys;
-  swarmOpts.dns = typeof swarmOpts.dns === 'undefined' ? false : swarmOpts.dns;
 
-  var port = swarmOpts.port || 0;
-  delete swarmOpts.port;
+  //swarmOpts.bootstrap = ['ws://hyperswarm.mauve.moe']
+  //swarmOpts.bootstrap = ['ws://localhost:4977']
```
Comment (staltz) [inline index.js:13]: Are these going to be uncommented? Is it required that we provide at least some bootstrap server? (If so, the README needs to be updated, and preferably, throw an error if it's missing)

File: index.js [hunk 3]
```diff
@@ -1,22 +1,18 @@
 var pull = require('pull-stream');
 var toPull = require('stream-to-pull-stream');
 var datDefaults = require('dat-swarm-defaults');
```
Comment (staltz) [inline index.js:3]: If this is not going to be used anymore, remove it from here as well as from the package.json

File: index.js [hunk 4]
```diff
@@ -172,7 +165,6 @@ module.exports = function makePlugin(opts) {
       });
 
       return () => {
-        clientPeer.close();
```
Comment (staltz) [inline index.js:175]: Does hyperswarm not provide any API to close? I think it's an important part

**Review by staltz — CHANGES_REQUESTED** [review]
Thank you Kyle for the PR. I haven't tested in CLI, but I noticed some changes needed in the code

File: index.js (hunk 1 above)
Comment (KyleMaas) [reply index.js:50]: Can do.

File: index.js (hunk 2 above)
Comment (KyleMaas) [reply index.js:13]: I don't think they need to be.  Ideally the user of multiserver-dht *should* provide them, but both Hyperswarm and Hyperswarm-Web provide default servers which are used if nothing else is provided.  Those could probably just be removed, or left in as suggestions.

File: index.js (hunk 3 above)
Comment (KyleMaas) [reply index.js:3]: Probably a good idea.

File: index.js (hunk 4 above)
Comment (KyleMaas) [reply index.js:175]: It does not.  At least nothing documented not recommended in examples, as far as I've been able to find.

Comment (staltz) [reply index.js:175]: It has `destroy` https://github.com/hyperswarm/hyperswarm/blob/924f56ae2781f1c605916e8e070f474ac7fff0f0/swarm.js#L302

In general it's a good idea to dig into source code if you can't find examples or docs. Mafintosh's code is very decently structured.

Comment (KyleMaas) [reply index.js:175]: I'm not generally a fan of using undocumented APIs, but if you know Mafintosh is consistent with them, that's fine.  See latest commit.

**Comment by KyleMaas** [comment]
Changes made in latest commit.  Still works in CLI.  Haven't retested in browser, but I expect that none of this would break anything there.

**Comment by KyleMaas** [comment]
Seems to still work in the browser as well.

**Comment by staltz** [comment]
Thank you @KyleMaas, I'll make a new breaking change version.

**Comment by KyleMaas** [comment]
Excellent.  Thank you!

**Comment by KyleMaas** [comment]
Is it possible you could update ssb-dht-invite to use this new version please?

**Pull request merged**